core_impl/src/main/java/io/opencensus/trace/internal/ConcurrentIntrusiveList.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2017, Google Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.trace.internal;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import io.opencensus.trace.internal.ConcurrentIntrusiveList.Element;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * An {@code ConcurrentIntrusiveList<T>} is a doubly-linked list where the link pointers are
+ * embedded in the elements. This makes insertion and removal into a known position constant time.
+ *
+ * <p>Elements must derive from the {@code Element<T extends Element<T>>} base class:
+ *
+ * <pre>{@code
+ * class MyClass implements Element<MyClass> {
```

Comment:
A simple example would be:

class Element<T extends Element<T>> {
  private T prev = null;
  private T next = null;
  public T getPrev() { return prev;}
  public void setPrev(T prev) { this.prev = prev; }
  public T getNext() { return next; }
  public void setNext(T next) { this.next = next; }
}


class IntrusiveList<T extends Element<T>> {
  private T head = null;
  public void add(T e) {
    assert (e.getNext() == null && e.getPrev() == null);
    if (head == null) {
      head = e;
    } else {
      e.setNext(head);
      head.setPrev(e);
      head = e;
    }
  }

  public T pop() {
    T e = head;
    head = head.getNext();
    e.setNext(null);
    head.setPrev(null);
    return e;
  }

  public void print() {
    for (T e = head; e != null; e = e.getNext()) {
      System.out.print(e + " ");
    }
    System.out.println("");
  }
}


class Int extends Element<Int> {
  public int x;

  public Int(int x) {
    this.x = x;
  }

  @Override
  public String toString() {
    return "" + x;
  }
}


public class Main {
  public static void main(String[] args) {
    IntrusiveList<Int> a = new IntrusiveList<Int>();
    a.print();
    a.add(new Int(5));
    a.add(new Int(6));
    a.print();
    System.out.println(a.pop());
    a.add(new Int(7));
    a.add(new Int(8));
    a.print();
  }
}